compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -139,7 +140,32 @@ impl<'thir, 'tcx> Visitor<'thir, 'tcx> for UnsafetyVisitor<'tcx> {
         }
     }
 
+    fn visit_pat(&mut self, pat: &Pat<'tcx>) {
+        use pattern::PatKind::*;
+
+        match *pat.kind {
+            AscribeUserType { .. } | Constant { .. } => {}
+            _ => {
+                if let ty::Adt(adt_def, _) = pat.ty.kind() {
+                    // binding to wildcard is okay since that never reads anything and stops double errors
+                    // with implict wildcard branches from `if let`s
+                    if adt_def.is_union() && !matches!(*pat.kind, Binding { .. } | Wild) {
+                        // pattern matching with a union and not doing something like v = Union { bar: 5 }
+                        self.requires_unsafe(pat.span, AccessToUnionField);
+                        return;
+                    }
+                }
+            }
+        }
+
+        visit::walk_pat(self, pat);
+    }
+
     fn visit_expr(&mut self, expr: &'thir Expr<'thir, 'tcx>) {
+        match expr.kind {
+            ExprKind::Field { .. } | ExprKind::VarRef { .. } | ExprKind::Scope { .. } => {}
+            _ => self.in_possible_lhs_union_assign = false,
```

Comment:
That sounds like a dangerous thing to do -- to have *all other* expressions reset the "union" flag. When new expressions are added, this could easily lead to subtle bugs.